lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -5691,13 +5716,64 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
     return RefType::get(bundleType.getElement(fieldIdx).type.getConstType(
                             bundleType.isConst() ||
                             bundleType.getElement(fieldIdx).type.isConst()),
-                        refType.getForceable());
+                        refType.getForceable(), refType.getLayer());
   }
 
   return emitInferRetTypeError(
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+LogicalResult RefCastOp::verify() {
+
+  SymbolRefAttr layer;
+  auto layerBlockOp =
+      dyn_cast<LayerBlockOp>((*this)->getBlock()->getParentOp());
+  if (layerBlockOp)
+    layer = layerBlockOp.getLayerName();
+
+  // The dest layer must be the same as the source layer or a parent of it.
+  SymbolRefAttr layerPointer = layer;
+  auto destLayer = getType().getLayer();
+  for (;;) {
+    if (layerPointer == destLayer)
+      break;
+
+    if (!layerPointer) {
+      if (!layer)
+        return emitOpError()
+               << "cannot cast to a layer from outside a layerblock";
+      auto diag = emitOpError()
+                  << "casts to a probe associated with layer '" << destLayer
+                  << "' from a layerblock associated with layer '" << layer
+                  << "'. The cast op must be in a layerblock associated with "
+                     "or a child of layer '"
+                  << destLayer << "'.";
+      return diag.attachNote(layerBlockOp.getLoc())
+             << "the layerblock was declared here";
+    }
+
+    if (layerPointer.getNestedReferences().empty()) {
+      layerPointer = {};
+      continue;
+    }
+    layerPointer =
+        SymbolRefAttr::get(layerPointer.getRootReference(),
+                           layerPointer.getNestedReferences().drop_back());
+  }
+
+  return success();
+}
+
+LogicalResult RefCastOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto layer = getType().getLayer();
+  if (layer && !symbolTable.lookupSymbolIn(circuitOp, layer))
```

Comment:
```suggestion
  if (layer && !symbolTable.lookupSymbolIn<LayerOp>(circuitOp, layer))
```

Reply:
Fixed in https://github.com/llvm/circt/pull/6574/commits/68b4e01b15af105b51776e0225616d1cf8853e38.